ipa-core/src/secret_sharing/vector/transpose.rs

```diff
@@ -0,0 +1,800 @@
+        impl TransposeFrom<&[$src_row; $src_rows]> for [$dst_row; $src_cols] {
+            fn transpose_from(&mut self, src: &[$src_row; $src_rows]) {
+                do_transpose_16(
+                    $src_rows / 16,
+                    $src_cols / 16,
+                    |i, j| {
+                        let mut d = [0u8; 32];
+                        for k in 0..16 {
+                            d[2 * k..2 * (k + 1)].copy_from_slice(
+                                &src[16 * i + k].as_raw_slice()[2 * j..2 * (j + 1)],
+                            );
+                        }
+                        d
+                    },
+                    |i, j, d| {
+                        for k in 0..16 {
+                            self[16 * i + k].as_raw_mut_slice()[2 * j..2 * (j + 1)]
+                                .copy_from_slice(&d[2 * k..2 * (k + 1)]);
+                        }
+                    },
+                );
+            }
+        }
+
+        #[cfg(all(test, unit_test))]
+        #[test]
+        fn $test_fn() {
+            tests::test_transpose_ba_to_ba::<$dst_row, $src_row, $src_rows, $src_cols>();
+        }
+
+        impl TransposeFrom<&BitDecomposed<$src_row>> for Vec<$dst_row> {
+            fn transpose_from(&mut self, src: &BitDecomposed<$src_row>) {
+                self.resize($src_cols, <$dst_row>::ZERO);
+                let src = <&[$src_row; $src_rows]>::try_from(&**src).unwrap();
+                let dst = <&mut [$dst_row; $src_cols]>::try_from(&mut **self).unwrap();
+                dst.transpose_from(src);
+            }
+        }
+
+        impl TransposeFrom<&[$src_row; $src_rows]> for Vec<$dst_row> {
+            fn transpose_from(&mut self, src: &[$src_row; $src_rows]) {
+                self.resize($src_cols, <$dst_row>::ZERO);
+                let dst = <&mut [$dst_row; $src_cols]>::try_from(&mut **self).unwrap();
+                dst.transpose_from(src);
+            }
+        }
+    };
+}
+
+impl_transpose_ba_to_ba!(BA16, BA64, 16, 64, test_transpose_ba_16x64);
+impl_transpose_ba_to_ba!(BA64, BA64, 64, 64, test_transpose_ba_64x64);
+impl_transpose_ba_to_ba!(BA256, BA64, 256, 64, test_transpose_ba_256x64);
+impl_transpose_ba_to_ba!(BA256, BA256, 256, 256, test_transpose_ba_256x256);
+
+/// Implement a transpose of a MxN matrix of secret-shared bits represented as
+/// `[AdditiveShare<Boolean, N>; <M>]` into a NxM bit matrix represented as `[AdditiveShare<BA<M>>; N]`.
+///
+/// For MxN = 256x64, the invocation looks like `impl_transpose_bool_to_ba!(BA256, 256, 64)`.
+macro_rules! impl_transpose_shares_bool_to_ba {
+    ($dst_row:ty, $src_rows:expr, $src_cols:expr, $test_fn:ident) => {
+        impl TransposeFrom<&[AdditiveShare<Boolean, $src_cols>; $src_rows]>
+            for [AdditiveShare<$dst_row>; $src_cols]
+        {
+            fn transpose_from(&mut self, src: &[AdditiveShare<Boolean, $src_cols>; $src_rows]) {
+                // Transpose left share
```

Comment:
there might be some opportunities for reducing code duplication here, although I am not sure if it is going to make the code look more complex or no

Reply:
I agree, it is repetitive. I'm not sure the best strategy to reduce duplication. Structuring the closures to read a single element rather than a block would make it less repetitive, but the generated code was substantially worse when I experimented with that.

There might be something with macros if the hygiene rules don't get in the way.

I don't think there will be too much need to modify this code (adding new transposes should mostly be possible by adding macro invocations, rather than by editing this code), but if you feel strongly that it's worth trying to make this less repetitive, I can work on it.